linalg/solvers.hpp

```diff
@@ -163,7 +163,7 @@ class IterativeSolver : public Solver
 
    mutable int final_iter;
    mutable bool converged;
-   mutable double final_norm;
+   mutable double initial_norm, final_norm;
```

Comment:
What do you think about initializing these to some known "placeholder" value so that e.g. `GetInitialNorm` won't return uninitialized data if the derived class doesn't set `initial_norm`?

In some sense this PR is adding the API requirement that setting `initial_norm` is a requirement for classes inheriting from `IterativeSolver`.

Reply:
I thought about this, but left the same behavior as the other fields. They pose the same requirement. If we add default values for all, that changes behavior of existing code. I'm fine adding a default for initial_norm (like `-1`?), especially that some external solvers may not provide the initial norm, but it will look weird compared to the other fields. Open to suggestions.

Reply:
I don't think there's an issue adding a default value of -1 for the other values as well: nobody should be depending on getting back uninitialized data by accessing these values before calling `Mult`, so I don't think that's a behavior we need to worry about changing.

Reply:
But maybe also if you do use -1, then `final_norm / initial_norm` should be replaced by `final_norm / std::abs(initial_norm)`, so the case of `-1 / -1` isn't misleading.

Reply:
I updated it, take a look now.